### DISCUSSION TITLE
我们想把idraw的渲染层，换成pixijs的实现，是否有可行性呢？

### DISCUSSION THREAD

**Discussion by xeoshow** [idrawjs/idraw#295] [General]
pixjs的性能会好很多，pixijs也有直接的UI项目，封装了很多组件，也支持动态元素比如gif和视频等，是我们项目需要的。是否主要就是下面这部分？谢谢。
https://github.com/idrawjs/idraw/tree/main/packages/renderer/src/draw

**Comment by chenshenhai** [comment]
@xeoshow 理论上是可以替换，但是需要考虑缩放过程中，要重新计算所有元素内容，保证无损的矢量渲染效果。

**Reply by xeoshow** [reply]
因为刚接触idraw不太了解哈，idraw现有的逻辑是直接操作canvas的吗？目前保证无损的矢量渲染处理，大概是在哪个代码部分呢？如果我们这边整合成功，后面需要的话，我们也可以提交pr。
非常感谢。

**Reply by chenshenhai** [reply]
> 因为刚接触idraw不太了解哈，idraw现有的逻辑是直接操作canvas的吗？目前保证无损的矢量渲染处理，大概是在哪个代码部分呢？如果我们这边整合成功，后面需要的话，我们也可以提交pr。
> 非常感谢。

@xeoshow  现有逻辑是间接操作Canvas，封装了一个Context2D的类来统一控制devicePixelRatio的变换，从而抹平devicePixelRatio的像素差异。https://github.com/idrawjs/idraw/blob/main/packages/util/src/lib/context2d.ts

矢量渲染主要是对每种元素进行尺寸特殊计算。比如这个矩形背景的计算 https://github.com/idrawjs/idraw/blob/main/packages/renderer/src/draw/box.ts#L99  利用通用视图缩放偏移信息`ViewScaleInfo`和视图尺寸信息`ViewSizeInfo`进行换算，换算出显示最终视图。

**Reply by xeoshow** [reply]
非常感谢答复。那对于比如图片这样的呢？因为图片本身并不是矢量化的。
pixijs应该是有专门的消除锯齿的filter，来应对类似情况。

**Reply by chenshenhai** [reply]
@xeoshow 图片保持位图的渲染，其它元素比如矩形、圆形、SVG等就保持矢量渲染。

**Reply by xeoshow** [reply]
你们在哪个城市呢？看看有没有进一步合作的可能，方便的话可以加个微信

**Reply by chenshenhai** [reply]
@xeoshow 技术交流直接email沟通就好，邮箱地址在我的github主页

**Reply by xeoshow** [reply]
已发邮件

**Comment by chenshenhai** [comment]
`iDraw.js`目前是一个实验型的学习项目，尽量会用原生Canvas来进行开发，暂时不考虑引入其它第三方绘图库或者框架。
